Add discovery notify support and mysensors notify

MartinHjelmare opened the pull request:
**Description:**
* Add mysensors notify platform
  * Make add_devices optional in platform callback function.
  * Use new argument structure for all existing mysensors platforms.
  * Add notify platform.
  * Update mysensors gateway.
* Refactor notify setup
  * Enable discovery of notify platforms.
  * Update and add tests for notify component and some platforms.
  * Continue setup of notify platforms if a platform fails setup.
  * Add and change tests that check platform config to use voluptuous config validation.
  * Add config validation to APNS notify platform.
  * **BREAKING CHANGE: Move APNS service under notify domain.**
  * Use discovery to set up mysensors notify platform.

**Related issue (if applicable):**
Previous PR: #4014 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):**
https://github.com/home-assistant/home-assistant.github.io/pull/1736
https://github.com/home-assistant/home-assistant.github.io/pull/1772

**Checklist:**
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

MartinHjelmare committed:
Add mysensors notify platform

* Make add_devices optional in platform callback function.
* Use new argument structure for all existing mysensors platforms.
* Add notify platform.
* Update mysensors gateway.

MartinHjelmare committed:
Refactor notify setup

* Enable discovery of notify platforms.
* Update and add tests for notify component and some platforms.
* Continue setup of notify platforms if a platform fails setup.
* Remove notify tests that check platform config. These tests are not
  needed when config validation is used.
* Add config validation to APNS notify platform.
* Use discovery to set up mysensors notify platform.

balloob commented on homeassistant/components/notify/__init__.py:
```diff
@@ -66,27 +66,31 @@ def send_message(hass, message, title=None, data=None):
+        if p_config is None:
+            p_config = {}
+        if discovery_info is not None:
+            p_config = discovery_info
```
I think that this is the wrong approach. This way a notify platform is unable to distinguish between discovery data (which cannot be trusted) and config from the user. Especially dangerous for `notify.command_line` !

MartinHjelmare replied:
Ok, I'll add a keyword argument to the `get_service` function signature instead. I'll have to update all platforms.

robbiet480 commented:
😮 discovery support in notify is going to seriously help the iOS app... awesome job!!!

pvizeli commented:
I wait until this PR is merge to convert notify to async.

MartinHjelmare committed:
Add discovery_info to get_service and update tests

* Add discovery_info as keyword argument to the get_service function
  signature and update all notify platforms.
* Update existing notify tests to check config validation using test
  helper.
* Add removed tests back in that checked config in apns, command_line
  and file platforms, but use config validation test helper to verify
  config.
* Add a test for notify file to increase coverage.
* Fix some PEP issues.

balloob commented on homeassistant/components/mysensors.py:
```diff
@@ -187,7 +190,10 @@ def mysensors_callback(gateway, node_id):
                 if key in devices:
-                    devices[key].update_ha_state(True)
+                    if add_devices:
+                        devices[key].update_ha_state(True)
```
`schedule_update_ha_state()`

balloob commented on homeassistant/components/mysensors.py:
```diff
@@ -197,11 +203,12 @@ def mysensors_callback(gateway, node_id):
+                if add_devices:
+                    _LOGGER.info('Adding new devices: %s', devices[key])
+                    add_devices([devices[key]])
                     devices[key].update_ha_state(True)
```
`schedule_update_ha_state()`

balloob commented on homeassistant/components/notify/apns.py:
```diff
@@ -12,17 +12,27 @@
+PLATFORM_SCHEMA = vol.Schema({
+    vol.Required(CONF_PLATFORM): cv.string,
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_CERTFILE): cv.string,
```
`cv.file` ?

balloob commented on homeassistant/components/notify/apns.py:
```diff
@@ -12,17 +12,27 @@
 ATTR_NAME = "name"
 
+PLATFORM_SCHEMA = vol.Schema({
+    vol.Required(CONF_PLATFORM): cv.string,
```
We can hardcode this to 'apns'

balloob commented on homeassistant/components/notify/apns.py:
```diff
@@ -12,17 +12,27 @@
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers import template as template_helper
 
 DOMAIN = "apns"
```
Platforms don't have a domain

MartinHjelmare replied:
This was used to register a service under "apns". See https://home-assistant.io/components/notify.apns/. It doesn't seem to be normal procedure. In the docs it says that it will register two services but I only see one registration in the platform code. Not really sure how this is supposed to work. I'll have to read some more.

balloob commented on homeassistant/components/notify/apns.py:
```diff
@@ -12,17 +12,27 @@
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_CERTFILE): cv.string,
+    vol.Required(CONF_TOPIC): cv.string,
+}, extra=vol.ALLOW_EXTRA)
```
Why allow extra?

MartinHjelmare replied:
Probably not needed, I copied it from `helpers/config_validation.py`.

balloob commented on homeassistant/components/notify/mysensors.py:
```diff
@@ -0,0 +1,65 @@
+from homeassistant.components import mysensors
+from homeassistant.components.notify import (ATTR_TARGET,
+                                             BaseNotificationService)
+
```
Add `DEPENDENCIES = ['mysensors']` ?

MartinHjelmare replied:
This will only be set up via discovery from the mysensors component. So I don't think the platform needs to depend explicitly on the mysensors component. If I add it will it have any consequences for the normal setup via discovery? No other mysensors platforms have this set.

balloob commented on homeassistant/components/notify/mysensors.py:
```diff
@@ -0,0 +1,65 @@
+    def send_message(self, message="", **kwargs):
+        """Send a message to a user."""
+        target_devices = kwargs.get(ATTR_TARGET)
+        devices = [
```
I think that list comprehensions like these are confusing to figure out what is going on, prefer written out.

balloob commented:
Awesome work!

MartinHjelmare commented on tests/components/notify/test_demo.py:
```diff
@@ -78,12 +80,14 @@ def test_no_notify_service(self, mock_demo_get_service):
+        assert notify.DOMAIN not in self.hass.config.components
         discovery.load_platform(
-            self.hass, 'notify', 'demo', {'test_key:test_val'}, CONFIG)
+            self.hass, 'notify', 'demo', {'test_key:test_val'}, {})
```
Need to fix this: I've made a typo in the discovery info so it's not a dictionary, but a set.

MartinHjelmare committed:
Fix comments and use more constants

MartinHjelmare committed:
Move apns notify service under notify domain
